### PR TITLE
[Enhancement] Support overflow mode for decimal type

### PR DESCRIPTION
**Pull request by liuyehcf**
## Main work

Support error reporting when decimal overflow by introducing a OverflowMode, which has three different modes:
* IGNORE
* OUTPUT_NULL
* REPORT_ERROR

This behavior can be enabled by setting session varaible `set sql_mode='ERROR_IF_OVERFLOW'`

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 66 / 109 (60.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|14|00.00%|[64, 78, 80, 83, 85, 87, 88, 92, 94, 98, 102, 187, 218, 223]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|8|25|32.00%|[71, 77, 83, 89, 114, 115, 138, 139, 141, 142, 143, 145, 146, 160, 180, 204, 205]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|19|42.11%|[1086, 1088, 1109, 1110, 1111, 1112, 1113, 1114, 1119, 1121, 1289]|
|:large_blue_circle:|src/exprs/math_functions.h|1|2|50.00%|[377]|
|:large_blue_circle:|src/exprs/function_context.cpp|4|4|100.00%|[]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|44|44|100.00%|[]|
|:large_blue_circle:|src/runtime/runtime_state.h|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 66 / 109 (60.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|14|00.00%|[64, 78, 80, 83, 85, 87, 88, 92, 94, 98, 102, 187, 218, 223]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|8|25|32.00%|[71, 77, 83, 89, 114, 115, 138, 139, 141, 142, 143, 145, 146, 160, 180, 204, 205]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|19|42.11%|[1086, 1088, 1109, 1110, 1111, 1112, 1113, 1114, 1119, 1121, 1289]|
|:large_blue_circle:|src/exprs/math_functions.h|1|2|50.00%|[377]|
|:large_blue_circle:|src/exprs/function_context.cpp|4|4|100.00%|[]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|44|44|100.00%|[]|
|:large_blue_circle:|src/runtime/runtime_state.h|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 2 / 3 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|3|66.67%|[2527]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 66 / 109 (60.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|14|00.00%|[64, 78, 80, 83, 85, 87, 88, 92, 94, 98, 102, 187, 218, 223]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|8|25|32.00%|[71, 77, 83, 89, 114, 115, 138, 139, 141, 142, 143, 145, 146, 160, 180, 204, 205]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|19|42.11%|[1086, 1088, 1109, 1110, 1111, 1112, 1113, 1114, 1119, 1121, 1289]|
|:large_blue_circle:|src/exprs/math_functions.h|1|2|50.00%|[377]|
|:large_blue_circle:|src/exprs/function_context.cpp|4|4|100.00%|[]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|44|44|100.00%|[]|
|:large_blue_circle:|src/runtime/runtime_state.h|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 2 / 3 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|3|66.67%|[2527]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 66 / 109 (60.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|14|00.00%|[64, 78, 80, 83, 85, 87, 88, 92, 94, 98, 102, 187, 218, 223]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|8|25|32.00%|[71, 77, 83, 89, 114, 115, 138, 139, 141, 142, 143, 145, 146, 160, 180, 204, 205]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|19|42.11%|[1086, 1088, 1109, 1110, 1111, 1112, 1113, 1114, 1119, 1121, 1289]|
|:large_blue_circle:|src/exprs/math_functions.h|1|2|50.00%|[377]|
|:large_blue_circle:|src/exprs/function_context.cpp|4|4|100.00%|[]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|44|44|100.00%|[]|
|:large_blue_circle:|src/runtime/runtime_state.h|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 2 / 3 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|3|66.67%|[2520]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 66 / 109 (60.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|14|00.00%|[64, 78, 80, 83, 85, 87, 88, 92, 94, 98, 102, 187, 218, 223]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|8|25|32.00%|[71, 77, 83, 89, 114, 115, 138, 139, 141, 142, 143, 145, 146, 160, 180, 204, 205]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|19|42.11%|[1086, 1088, 1109, 1110, 1111, 1112, 1113, 1114, 1119, 1121, 1289]|
|:large_blue_circle:|src/exprs/math_functions.h|1|2|50.00%|[377]|
|:large_blue_circle:|src/exprs/function_context.cpp|4|4|100.00%|[]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|44|44|100.00%|[]|
|:large_blue_circle:|src/runtime/runtime_state.h|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 54 / 104 (51.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/expr_context.cpp|0|2|00.00%|[189, 190]|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|15|00.00%|[61, 78, 79, 80, 82, 83, 86, 87, 88, 90, 91, 92, 97, 100, 101]|
|:large_blue_circle:|src/exprs/function_context.cpp|0|2|00.00%|[159, 160]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|2|00.00%|[342, 343]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|0|12|00.00%|[109, 111, 114, 157, 158, 160, 174, 176, 179, 199, 201, 204]|
|:large_blue_circle:|src/exprs/math_functions.h|0|3|00.00%|[376, 378, 381]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|16|50.00%|[1086, 1087, 1088, 1091, 1102, 1110, 1118, 1285]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|46|52|88.46%|[75, 94, 112, 212, 319, 366]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 2 / 3 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|3|66.67%|[2520]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 2 / 3 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|3|66.67%|[2520]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 54 / 104 (51.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/expr_context.cpp|0|2|00.00%|[189, 190]|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|15|00.00%|[61, 78, 79, 80, 82, 83, 86, 87, 88, 90, 91, 92, 97, 100, 101]|
|:large_blue_circle:|src/exprs/function_context.cpp|0|2|00.00%|[159, 160]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|2|00.00%|[342, 343]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|0|12|00.00%|[109, 111, 114, 157, 158, 160, 174, 176, 179, 199, 201, 204]|
|:large_blue_circle:|src/exprs/math_functions.h|0|3|00.00%|[376, 378, 381]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|16|50.00%|[1086, 1087, 1088, 1091, 1102, 1110, 1118, 1285]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|46|52|88.46%|[75, 94, 112, 212, 319, 366]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SqlModeHelper.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 54 / 104 (51.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/expr_context.cpp|0|2|00.00%|[189, 190]|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|15|00.00%|[61, 78, 79, 80, 82, 83, 86, 87, 88, 90, 91, 92, 97, 100, 101]|
|:large_blue_circle:|src/exprs/function_context.cpp|0|2|00.00%|[159, 160]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|2|00.00%|[342, 343]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|0|12|00.00%|[109, 111, 114, 157, 158, 160, 174, 176, 179, 199, 201, 204]|
|:large_blue_circle:|src/exprs/math_functions.h|0|3|00.00%|[376, 378, 381]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|16|50.00%|[1086, 1087, 1088, 1091, 1102, 1110, 1118, 1285]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|46|52|88.46%|[75, 94, 112, 212, 319, 366]|

**Review by satanson — DISMISSED** [review]

File: test/sql/test_decimal/R/test_decimal_overflow [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+-- result:
+None
+-- !result
+select cast(c_d128 * 1.000000000 as decimal128) from t_decimal_overflow where c_id = 2;
+-- result:
+None
+-- !result
+select /*+ SET_VAR(sql_mode='ERROR_IF_OVERFLOW')*/ cast(c_d32 * c_d32 as decimal32) from t_decimal_overflow where c_id = 1;
+-- result:
+E: (1064, 'Expr evaluate meet error: Decimal overflow')
```
Comment (fzhedu) [inline test/sql/test_decimal/R/test_decimal_overflow:78]: from the error, we can't know which expr report this error, can the error specify the operator?

Comment (liuyehcf) [reply test/sql/test_decimal/R/test_decimal_overflow]: Add more precise information.

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 54 / 112 (48.21%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/expr_context.cpp|0|2|00.00%|[189, 190]|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|17|00.00%|[62, 68, 71, 84, 85, 86, 88, 89, 92, 93, 94, 96, 97, 98, 103, 107, 108]|
|:large_blue_circle:|src/exprs/function_context.cpp|0|2|00.00%|[159, 160]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|2|00.00%|[342, 343]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|0|12|00.00%|[109, 111, 114, 157, 158, 160, 174, 176, 179, 199, 201, 204]|
|:large_blue_circle:|src/exprs/math_functions.h|0|3|00.00%|[376, 378, 381]|
|:large_blue_circle:|src/exprs/arithmetic_operation.h|0|6|00.00%|[81, 83, 85, 91, 93, 97]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|16|50.00%|[1086, 1087, 1088, 1091, 1102, 1110, 1118, 1285]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|46|52|88.46%|[75, 94, 112, 212, 319, 366]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SqlModeHelper.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 54 / 112 (48.21%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/expr_context.cpp|0|2|00.00%|[189, 190]|
|:large_blue_circle:|src/exprs/decimal_binary_function.h|0|17|00.00%|[62, 68, 71, 84, 85, 86, 88, 89, 92, 93, 94, 96, 97, 98, 103, 107, 108]|
|:large_blue_circle:|src/exprs/function_context.cpp|0|2|00.00%|[159, 160]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|2|00.00%|[342, 343]|
|:large_blue_circle:|src/exprs/arithmetic_expr.cpp|0|12|00.00%|[109, 111, 114, 157, 158, 160, 174, 176, 179, 199, 201, 204]|
|:large_blue_circle:|src/exprs/math_functions.h|0|3|00.00%|[376, 378, 381]|
|:large_blue_circle:|src/exprs/arithmetic_operation.h|0|6|00.00%|[81, 83, 85, 91, 93, 97]|
|:large_blue_circle:|src/exprs/cast_expr.cpp|8|16|50.00%|[1086, 1087, 1088, 1091, 1102, 1110, 1118, 1285]|
|:large_blue_circle:|src/exprs/decimal_cast_expr.h|46|52|88.46%|[75, 94, 112, 212, 319, 366]|

**Review by satanson — APPROVED** [review]

**Review by fzhedu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**